### ISSUE TITLE
Can I unit-test my custom rules?

### ISSUE THREAD

**Issue by vitaly-magyari** [stoplightio/spectral#1292]
**User story.**
As a custom ruleset user, I can write unit-test that can run my custom rules, so that I can verify their behavior.

**Is your feature request related to a problem?**
I can't test my custom rules at all, except for running entire ruleset against a spec. And can't use things like jest for them.

**Describe the solution you'd like**
In the base repo here there are tests that create spectral instance with `new Spectral()`. Is anything of the sort possible outside? Any possible alternatives?

**Comment by adanilev** [comment]
Hi @vitaly-magyari, here's an example of a unit test that targets a single custom rule. I use jest. IIRC, the structure was ~~ripped off~~ inspired by tests in this repo:

```ts
import { Spectral } from "@stoplight/spectral";
import * as ruleset from "../../contact-object.json";
import cloneDeep = require("lodash.clonedeep");

const sut = "contact-email-defined";

describe(sut, () => {
  const spectral = new Spectral();
  spectral.setRules({ [sut]: cloneDeep(ruleset.rules[sut] as any) });

  it("should not return any results for a valid input", async () => {
    const res = await spectral.run({
      openapi: "3.0.0",
      info: { contact: { email: "here@invalid.com" } }
    });

    expect(res.length).toEqual(0);
  });

  it("should return an error if the email is missing", async () => {
    const res = await spectral.run({
      openapi: "3.0.0",
      info: { contact: { name: "not an email" } }
    });

    expect(res.length).toEqual(1);
    expect(res).toEqual([
      {
        code: "contact-email-defined",
        message: "The contact object must include an email address",
        path: ["info", "contact"],
        range: {
          end: {
            character: 28,
            line: 4
          },
          start: {
            character: 14,
            line: 3
          }
        },
        severity: 0
      }
    ]);
  });
});
```

**Comment by P0lip** [comment]
Hey @vitaly-magyari 
Was the example above sufficient? It's a nice example.
Feel free to re-open if you need further assistance.

Thanks for putting this together @adanilev!

**Comment by MarcusAckermann** [comment]
A common mistake is to provide OpenAPI documents with tests which are not valid as far as the schema is concerned. I would like to enable 'oas3-schema' along with the rule to be tested, but I do not have a clue how to configure spectral in the test example above. Can anybody help?
